Update Product and Services.cpp

nniricha opened the pull request.

nniricha committed:
Update Product and Services.cpp

VideshBridgelal commented on Client Database System/Client Database System/Product and Services.cpp:
```diff
@@ -8,13 +8,74 @@
+void Product_and_Services::AddNewService(string product, string serviceDescription, vector<string>& productVector, vector<string>& descriptionVector){
+    cout << "Input product, service description: ";
+    cin >> product;
+    cin >> serviceDescription;
```
Using cin will only read one word. The fix for this would be `getline(cin, serviceDescription);` to get the entire description

VideshBridgelal commented on Client Database System/Client Database System/Product and Services.cpp:
```diff
@@ -8,13 +8,74 @@
+    cin >> userService;
+    
+    unsigned int i;
+    for(i = 0; i < productVector.size(); ++i){
+        if(userProduct == productVector.at(i)){
+            cout << "Input new product: " << endl;
+            cin >> userProduct;
+        }
+    }
+    for(i = 0; i < descriptionVector.size(); ++i){
+        if(userService == descriptionVector.at(i)){
+            cout << "Input new service: " << endl;
+            cin >> userService;
+        }
```
This reads the users input but we have to add them to the vector, and we also have to erase the old elements. We can do this with the `insert()` and `erase()` functions.

So, in this case, once you find the product you can get the input and use `productVector.insert(productVector.begin() + i , userProduct);` 
The first part "productVector.begin()" tells the insert to put userProduct at element 0. "+ i" moves the index number from zero to whatever number "i" is in the loop. Then "userProduct" tells insert() to insert "userProduct" before the the index.

erase() would be written similarly to insert(). For the productVector, it would be `productVector.erase(productVector.begin() + i + 1);`
The only difference here is that it has no variable to input. Instead it finds the index that you want to modify. Like here `productVector.begin()` starts the index at 0, then `+ i + 1` changes the index number to whatever i is in the loop plus 1 because we want to erase the old element, which was pushed up by 1 from the insert function.\

VideshBridgelal commented on Client Database System/Client Database System/Product and Services.cpp:
```diff
@@ -8,13 +8,74 @@
     priceVector.push_back(userPrice);
 }
+
+void Product_and_Services::ReadFromFile(vector<string>& productVector, vector<string>& descriptionVector){
```
Yeah, these functions are quite confusing because we're dealing with external files.

So this requires the fstream library which has been included already. First what we do is create a reference to the input-stream by writing `ifstream inputStreamVariable;` This tells the program that inputStreamVariable is reading information from the file. But we have to tell the program where to get the information from.

To do this we write `inputStreamVariable.open("TextFile.txt");` and I already included a .txt file which is named "SaleRepresentativesFile.txt" so we would use that instead of "TextFile.txt". The open() function opens the file so that the program can read its contents.

Now we can read the inputs from the file, but we don't know how much information is in the file, so we use a while loop. And the argument is `(!inputStreamVariable.eof())` and `eof()` checks to see when it reaches the end of the file. Then, in the loop we get the input from the file and fill it into the vector.

VideshBridgelal replied:
And I forgot that we need to close the file before we finish. So after the loop we would write `inputStreamVariable.close()` to close the file from the program.

VideshBridgelal commented on Client Database System/Client Database System/Product and Services.cpp:
```diff
@@ -8,13 +8,74 @@
+    }
+}
+
+void Product_and_Services::SaveToFile(vector<string>& productVector, vector<string>& descriptionVector){
```
This one is similar to the ReadFromFile function.

This time we use `ofstream outputStreamVariable;` telling the program that we are using the variable outputStreamVariable to output information to the file.

To open the file we use `outputStreamVariable.open("TextFile.txt");` but we're still using the same text file as before, "SaleRepresentativesFile.txt".

Also, you should probably include the string variables product and serviceDescription in the function's arguments because we can't output to the file straight from the vector, at least it didn't work with me during my last project. So that would look like `product = productVector.at(i);` and then we start outputting information into the file using `outputStreamVariable << product;`. And that goes for the decriptionVector too.

After the loop is done we close the file with `outputStreamVariable.close();` and that's it, besides from making it look pretty in the file.

nniricha committed:
Update Product and Services.cpp

On the SaveToFile, I wasn't sure is product and service description have separate loops? I think thats my only problem left.

VideshBridgelal commented on Client Database System/Client Database System/Product and Services.cpp:
```diff
@@ -8,13 +8,109 @@
+        
+        prodAndServOut << serviceDescription;
+    }
+    prodAndServIn.close();
```
Has to be `prodAndServOut.close();`

VideshBridgelal commented on Client Database System/Client Database System/Product and Services.cpp:
```diff
@@ -8,13 +8,109 @@
+    ifstream prodAndServIn;
+    prodAndServIn.open("ProductsAndServicesFile.txt");
+    
+    while(!inputStreamVariable.eof()){
```
This was my fault, inputStreamVariable was supposed to be an example, but you can name it anything. In this case you named it prodAndServIn, so whenever you want to reference the .txt file you would use the defined variable.

nniricha committed:
Update Product and Services.cpp
